pinot-segment-local/src/main/java/org/apache/pinot/segment/local/recordtransformer/SchemaConformingTransformerV2.java

```diff
@@ -0,0 +1,715 @@
+            .anyMatch(suffix -> !kv.getKey().endsWith(suffix))).forEach(kv -> {
+      if (null == mergedTextIndexShinglingOverlapLength) {
+        generateTextIndexToken(kv, luceneTokens, mergedTextIndexTokenMaxLength);
+      } else {
+        generateShingleTextIndexToken(kv, luceneTokens, mergedTextIndexTokenMaxLength,
+            mergedTextIndexShinglingOverlapLength);
+      }
+    });
+    return luceneTokens;
+  }
+}
+
+/**
+ * SchemaTreeNode represents the tree node when we construct the schema tree. The node could be either leaf node or
+ * non-leaf node. Both types of node could hold the volumn as a column in the schema.
+ * For example, the schema with columns a, b, c, d.e, d.f, x.y, x.y.z, x.y.w will have the following tree structure:
+ * root -- a*
+ *      -- b*
+ *      -- c*
+ *      -- d -- e*
+ *           -- f*
+ *      -- x* -- y* -- z*
+ *                  -- w*
+ * where node with "*" could represent a valid column in the schema.
+ */
+class SchemaTreeNode {
+  private boolean _isColumn;
+  private Map<String, SchemaTreeNode> _children;
+  // Taking the example of key "x.y.z", the keyName will be "z" and the parentPath will be "x.y"
+  // Root node would have keyName as "" and parentPath as null
+  // Root node's children will have keyName as the first level key and parentPath as ""
+  @Nonnull
+  private String _keyName;
+  @Nullable
+  private String _columnName;
+  @Nullable
+  private String _parentPath;
+  private FieldSpec _fieldSpec;
+
+  public SchemaTreeNode(String keyName, String parentPath, Schema schema) {
+    _keyName = keyName;
+    _parentPath = parentPath;
+    _fieldSpec = schema.getFieldSpecFor(getJsonKeyPath());
+    _children = new HashMap<>();
+  }
+
+  public boolean isColumn() {
+    return _isColumn;
+  }
+
+  public void setColumn(String columnName) {
+    if (columnName == null) {
+      _columnName = getJsonKeyPath();
+    } else {
+      _columnName = columnName;
+    }
+    _isColumn = true;
+  }
+
+  public boolean hasChild(String key) {
+    return _children.containsKey(key);
+  }
+
+  /**
+   * If does not have the child node, add a child node to the current node and return the child node.
+   * If the child node already exists, return the existing child node.
+   * @param key
+   * @return
+   */
+  public SchemaTreeNode getAndCreateChild(String key, Schema schema) {
+    SchemaTreeNode child = _children.get(key);
+    if (child == null) {
+      child = new SchemaTreeNode(key, getJsonKeyPath(), schema);
+      _children.put(key, child);
+    }
+    return child;
+  }
+
+  public SchemaTreeNode getChild(String key) {
+    return _children.get(key);
+  }
+
+  public String getKeyName() {
+    return _keyName;
+  }
+
+  public String getColumnName() {
+    return _columnName;
+  }
+
+  public Object getValue(Object value) {
+    // If the field is a single value string field, but value is an array of collection, serialize it
```

Comment:
Can we also add some more details regarding the problem of letting downstream `DataTypeTransformer` doing the serialization? Specifically, it treats arrays as multi-variable "column" rather than serializing it to string despite the column type being string typed single-variable column. That is why we want to serialize objects here.